File: app/api/openagi/route.py
from fastapi import APIRouter, Depends, HTTPException

HTTPException
from fastapi.security import OAuth2PasswordBearer
from typing import Dict, Any
from sqlalchemy.orm import Session
from logging_config import logger

from database import get_db
from models import User
from openagi_integration import OpenAGI

router = APIRouter()
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

async def get_current_user(token: str = Depends(oauth2_scheme), db: Session = Depends(get_db)):
    from app import get_current_user as app_get_current_user
    return await app_get_current_user(token, db)

@router.post("/openagi")
async def process_openagi_request(
    request_data: Dict[str, Any],
    current_user: User = Depends(get_current_user)
):
    try:
        # Extract config and input from request
        config = request_data.get("config", {})
        input_text = request_data.get("input", "")
        history = request_data.get("history", [])
        
        # Initialize OpenAGI with the provided configuration
        openagi = OpenAGI(config)
        
        # Process the input
        response = openagi.process(input_text, history)
        
        return {
            "response": response,
            "config": config
        }
    except Exception as e:
        logger.error(f"Error processing OpenAGI request: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to process OpenAGI request: {str(e)}")

File: app/api/openagi/route.ts
import { type NextRequest, NextResponse } from "next/server"
import { z } from "zod"
import { getServerSession } from "next-auth/next"
import { authOptions } from "@/lib/auth"

// Define schema for OpenAGI request validation
const OpenAGIRequestSchema = z.object({
  config: z.record(z.any()),
  input: z.string(),
})

export async function POST(request: NextRequest) {
  try {
    // Check authentication
    const session = await getServerSession(authOptions)

    if (!session || !session.user) {
      return NextResponse.json({ error: "Unauthorized" }, { status: 401 })
    }

    const body = await request.json()

    // Validate the request
    const result = OpenAGIRequestSchema.safeParse(body)

    if (!result.success) {
      return NextResponse.json({ error: "Invalid request structure", details: result.error.format() }, { status: 400 })
    }

    const { config, input } = result.data

    // In a real implementation, this would:
    // 1. Initialize the OpenAGI framework with the provided configuration
    // 2. Process the input through the OpenAGI pipeline
    // 3. Return the response

    // For now, simulate a response
    const response = simulateOpenAGIResponse(config, input)

    return NextResponse.json({
      response,
      config: config,
    })
  } catch (error) {
    console.error("Error processing OpenAGI request:", error)
    return NextResponse.json(
      { error: "Failed to process OpenAGI request", details: error instanceof Error ? error.message : "Unknown error" },
      { status: 500 },
    )
  }
}

// Simulate an OpenAGI response
function simulateOpenAGIResponse(config: any, input: string): string {
  // Extract agent type
  const agentType = config.agent?.type || "conversational-agent"

  // Extract tools
  const tools = config.tools || []

  // Generate a response based on the agent type and tools
  let response = `OpenAGI Agent (${agentType}) processing: "${input}"\n\n`

  if (tools.length > 0) {
    response += "Using the following tools:\n"
    tools.forEach((tool: any, index: number) => {
      response += `${index + 1}. ${tool.type}\n`
    })
    response += "\n"
  }

  response += "In a production environment, this would be processed by the actual OpenAGI framework."

  return response
}

